## Relevant RAD-seq related papers

- Miller, Michael R., Joseph P. Dunham, Angel Amores, William A. Cresko, and Eric A. Johnson. 2007. “Rapid and Cost-Effective Polymorphism Identification and Genotyping Using Restriction Site Associated DNA (RAD) Markers.” Genome Research 17 (2): 240–48. https://doi.org/10.1101/gr.5681207.


- Baird, Nathan A., Paul D. Etter, Tressa S. Atwood, Mark C. Currey, Anthony L. Shiver, Zachary A. Lewis, Eric U. Selker, William A. Cresko, and Eric A. Johnson. 2008. “Rapid SNP Discovery and Genetic Mapping Using Sequenced RAD Markers.” PLOS ONE 3 (10): e3376. https://doi.org/10.1371/journal.pone.0003376.


- Hohenlohe, Paul A., Susan Bassham, Paul D. Etter, Nicholas Stiffler, Eric A. Johnson, and William A. Cresko. 2010. “Population Genomics of Parallel Adaptation in Threespine Stickleback Using Sequenced RAD Tags.” PLOS Genetics 6 (2): e1000862. https://doi.org/10.1371/journal.pgen.1000862.

- Emerson, Kevin J., Clayton R. Merz, Julian M. Catchen, Paul A. Hohenlohe, William A. Cresko, William E. Bradshaw, and Christina M. Holzapfel. 2010. “Resolving Postglacial Phylogeography Using High-Throughput Sequencing.” Proceedings of the National Academy of Sciences 107 (37): 16196–200. https://doi.org/10.1073/pnas.1006538107.


- Elshire, Robert J., Jeffrey C. Glaubitz, Qi Sun, Jesse A. Poland, Ken Kawamoto, Edward S. Buckler, and Sharon E. Mitchell. 2011. “A Robust, Simple Genotyping-by-Sequencing (GBS) Approach for High Diversity Species.” PLOS ONE 6 (5): e19379. https://doi.org/10.1371/journal.pone.0019379.


- Wagner, Catherine E., Irene Keller, Samuel Wittwer, Oliver M. Selz, Salome Mwaiko, Lucie Greuter, Arjun Sivasundar, and Ole Seehausen. 2013. “Genome-Wide RAD Sequence Data Provide Unprecedented Resolution of Species Boundaries and Relationships in the Lake Victoria Cichlid Adaptive Radiation.” Molecular Ecology 22 (3): 787–98. https://doi.org/10.1111/mec.12023.


- Peterson, Brant K., Jesse N. Weber, Emily H. Kay, Heidi S. Fisher, and Hopi E. Hoekstra. 2012. “Double Digest RADseq: An Inexpensive Method for De Novo SNP Discovery and Genotyping in Model and Non-Model Species.” PLOS ONE 7 (5): e37135. https://doi.org/10.1371/journal.pone.0037135.


- Eaton, Deren A. R., and Richard H. Ree. 2013. “Inferring Phylogeny and Introgression Using RADseq Data: An Example from Flowering Plants (Pedicularis: Orobanchaceae).” Systematic Biology 62 (5): 689–706. https://doi.org/10.1093/sysbio/syt032.


- Eaton, Deren A. R. 2014. “PyRAD: Assembly of de Novo RADseq Loci for Phylogenetic Analyses.” Bioinformatics 30 (13): 1844–49. https://doi.org/10.1093/bioinformatics/btu121.


- Eaton, Deren A. R., Andrew L. Hipp, Antonio González-Rodríguez, and Jeannine Cavender-Bares. 2015. “Historical Introgression among the American Live Oaks and the Comparative Nature of Tests for Introgression.” Evolution 69 (10): 2587–2601. https://doi.org/10.1111/evo.12758.


- Andrews, Kimberly R., Jeffrey M. Good, Michael R. Miller, Gordon Luikart, and Paul A. Hohenlohe. 2016. “Harnessing the Power of RADseq for Ecological and Evolutionary Genomics.” Nature Reviews Genetics 17 (2): 81. https://doi.org/10.1038/nrg.2015.28.


- Glenn, Travis C., Roger Nilsen, Troy J. Kieran, John W. Finger, Todd W. Pierson, Kerin E. Bentley, Sandra Hoffberg, et al. 2016. “Adapterama I: Universal Stubs and Primers for Thousands of Dual-Indexed Illumina Libraries (ITru & INext).” BioRxiv, June, 049114. https://doi.org/10.1101/049114.

- Glenn, Travis C., Natalia J. Bayona-Vasquez, Troy J. Kieran, Todd W. Pierson, Sandra L. Hoffberg, Peter A. Scott, Kerin E. Bentley, et al. 2017. “Adapterama III: Quadruple-Indexed, Triple-Enzyme RADseq Libraries for about $1USD per Sample (3RAD).” BioRxiv, October, 205799. https://doi.org/10.1101/205799.


- Eaton, Deren A R, and Isaac Overcast. 2020. “Ipyrad: Interactive Assembly and Analysis of RADseq Datasets.” Edited by Russell Schwartz. Bioinformatics, January, btz966. https://doi.org/10.1093/bioinformatics/btz966.
