# RADCamp NYC 2019 Part II (Bioinformatics)
# Day 2 (AM)

## Overview of the morning activities:
* [Brief review of the 16 newly generated datasets](#overview-of-new-datasets)
* [Examine quality of the assemblies](#examine-quality-of-assemblies)
* [Next steps after you finish your first assembly](#next-steps-after-you-finish-your-first-assembly)
* [Intro to genome scale analysis](#intro-to-genome-scale-analysis)
* [ipyrad analysis API - Principal Component Analysis (PCA)](ipyrad-analysis-API---Principal-Component-Analysis-(PCA))
## Overview of new datasets
**put link to google drive slides here**
This is a 1-slide-maximum introduction to each of the new datasets generated
during part I of the workshop. Questions to answer include:
* What is the group?
* Geographic/taxonomic scope?
* What are the research questions? Are they more phylo- or population genetic scale.
* Do you have some ideas of the analysis you want to perform?`

## Examine quality of assemblies
Lead: Isaac

Lets take some time to look at results of the assembly process for some of the
real datasets. Did they all work perfectly? Why did some work, why did some
break?
* Look at runtimes: How long did they take to run?
* What do the stats files look like? Teaching how to read and interperet stats files 

## Next steps after you finish your first assembly
* Why would you want to re-run step 7 with different parameters?
* Talk about mindepth/minsamp. Branching.
* Filtering your data
* Missing data stuff
* Analysis

## Intro to genome scale analysis
Lead: Deren

Genomic analysis, why do we need more than one gene. What people do with radseq
data and why?
* Raxml advantages/shortcomings
* PCA advantages/shortcoming.
* SNPs vs gene trees?
* Shortcomings/advantages of RAD?
* Why are the analysis tools better than trying to go out on your own.

## ipyrad analysis API - Principal Component Analysis (PCA)


## ipyrad analysis API - Population STRUCTURE
Notebook (let it run over lunch)

