# RADCamp NYC 2023 Part II (Bioinformatics)
# Day 2 (AM)

## Overview of the morning activities:
* [Brief review of the newly generated 3RAD datasets](#overview-of-new-datasets)
* [Examine quality of the assemblies](#examine-quality-of-assemblies)
* [Next steps after you finish your first assembly](#next-steps-after-you-finish-your-first-assembly)
* [Intro to genome scale analysis](#intro-to-genome-scale-analysis)
* [Coffee Break](#coffee-break)
* [ipyrad analysis API - Principal Component Analysis (PCA)](#ipyrad-analysis-api-principal-component-analysis))
* [ipyrad analysis API - Population STRUCTURE](#ipyrad-analysis-api-Population-structure)

## Examine quality of assemblies

Lets take some time to look at results of the assembly process for some of the
real datasets. Did they all work perfectly? Why did some work, why did some
break?

[How did the runs proceed? (AKA What did Isaac do with his night last night)](PartII-Groups.txt)

[What the results of the demultiplexing process (step 1) should look like](Demux-Results.txt)


* Look at runtimes: How long did they take to run?
* What do the stats files look like? Teaching how to read and interperet stats files. 

## Next steps after you finish your first assembly

* Why would you want to re-run step 7 with different parameters?
* Talk about mindepth/minsamp. Branching.
* Filtering your data
* Missing data stuff
* Analysis

## Intro to genome scale analysis
Lead: Deren

Genomic analysis, why do we need more than one gene. What people do with radseq
data and why?
* Raxml advantages/shortcomings
* PCA advantages/shortcoming
* SNPs vs gene trees?
* Shortcomings/advantages of RAD?
* Why are the analysis tools better than trying to go out on your own.

## Coffee Break

## ipyrad analysis API Principal Component Analysis
Lead: Isaac


## ipyrad analysis API Population STRUCTURE


Notebook (let it run over lunch)

